File: server/apps/sync-service/src/main/java/com/sharetimer/syncservice/adapter/in/listener/TimerExpirationListener.java
package com.sharetimer.syncservice.adapter.in.listener;

import org.springframework.data.redis.connection.Message;
import org.springframework.data.redis.connection.MessageListener;
import org.springframework.data.redis.listener.PatternTopic;
import org.springframework.stereotype.Component;
import com.sharetimer.storage.redis.config.RedisMessageListenerContainerFactory;
import com.sharetimer.storage.redis.config.TimerRedisProps;
import com.sharetimer.syncservice.application.port.in.TimerUseCase;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Component
@RequiredArgsConstructor
@Slf4j
public class TimerExpirationListener implements MessageListener {

  private final TimerUseCase timerUseCase;
  private final TimerRedisProps timerRedisProps;
  private final RedisMessageListenerContainerFactory factory;

  @PostConstruct
  public void init() {
    int idx = timerRedisProps.getExpiration().getDbIndex();
    factory.getContainer(idx).addMessageListener(this,
        new PatternTopic(String.format("__keyevent@%d__:expired", idx)));
  }

  @Override
  public void onMessage(Message message, byte[] pattern) {
    String body = new String(message.getBody());
    log.info("Expired key: {}", body);

    if (body.startsWith(timerRedisProps.getExpiration().getKeyPrefix())) {
      String timerId = body.substring(timerRedisProps.getExpiration().getKeyPrefix().length());
      timerUseCase.processTimerExpiration(timerId);
    }
  }
}

File: server/apps/sync-service/src/main/kotlin/com/sharetimer/syncservice/adapter/input/listener/TimerExpirationListener.kt
package com.sharetimer.syncservice.adapter.input.listener

import com.sharetimer.storage.redis.config.RedisMessageListenerContainerFactory
import com.sharetimer.storage.redis.config.TimerRedisProps
import com.sharetimer.syncservice.application.port.input.TimerUseCase
import jakarta.annotation.PostConstruct
import org.slf4j.LoggerFactory
import org.springframework.data.redis.connection.Message
import org.springframework.data.redis.connection.MessageListener
import org.springframework.data.redis.listener.PatternTopic
import org.springframework.stereotype.Component

@Component
class TimerExpirationListener(
    private val timerUseCase: TimerUseCase,
    private val timerRedisProps: TimerRedisProps,
    private val factory: RedisMessageListenerContainerFactory,
) : MessageListener {
    @PostConstruct
    fun init() {
        factory.container.addMessageListener(this, PatternTopic("__keyevent@0__:expired"))
    }

    override fun onMessage(
        message: Message,
        pattern: ByteArray?,
    ) {
        val body = String(message.body)
        log.info("Expired key: {}", body)

        body
            .removePrefix(timerRedisProps.expiration.keyPrefix)
            .takeIf { it != body }
            ?.let { timerUseCase.processTimerExpiration(it) }
    }

    companion object {
        private val log = LoggerFactory.getLogger(TimerExpirationListener::class.java)
    }
}
